dimod/higherorder.py

```diff
@@ -200,21 +283,68 @@ def _prod(iterable):
     return val
 
 
+def _prod_D(iterable, D):
+    val = [1.] * D
+    for v in iterable:
+        val *= v
+    return val
+
+
+def _shifted_poly(h, j, label_dict):
+    poly = {}
+    for k, v in h.items():
+        new_tup = (label_dict[k],)
+        poly[new_tup] = v
+    for k, v in j.items():
+        new_tup = tuple(sorted((label_dict[vidx] for vidx in list(k))))
+        poly[new_tup] = v
+    return poly
+
+
+def create_poly(h, J):
+    """ given h,J creates a single polynomial dict.
+    all h's will turn into tuples.
+
+    Args:
+        h (dict): a dict of linear variables
+        J (dict): a dict of quadratic and higher order variables
+
+    Returns
+        dict: a higher order problem dict containing both linear,
+    quadratic and n-order terms.
+
+    """
+
+    poly = {(k,): v for k, v in h.items()}
+    poly.update(J)
+    return poly
+
+
 def poly_energy(sample, poly):
     """Create a binary quadratic model from a higher order polynomial.
 
     Args:
-        sample (dict):
-            Sample for which to calculate the energy, formatted as a dict where keys
-            are variables and values are the value associated with each variable.
+        sample (dict or (list,np.array)):
+            Sample(s) for which to calculate the energy, formatted as a dict 
+            where keys are variables and values are the value associated with 
+            each variable. If formatted as a list or np.array, the terms in 
+            poly dict must be provided as variable order.
 
         poly (dict):
-            Polynomial as a dict of form {term: bias, ...}, where `term` is a tuple of
-            variables and `bias` the associated bias.
+            Polynomial as a dict of form {term: bias, ...}, where `term` is a 
+            tuple of variables and `bias` the associated bias.
 
     Returns:
-        float: The energy of the sample.
+        float or list: The energy of the sample(s).
```

Comment:
`float/list` to be picked up correctly by readthedocs. You could also consider using https://github.com/dwavesystems/dimod/blob/96ef78f53b7762782491bdd635a2c2dd1d6c75d2/dimod/sampleset.py#L39 to extend to samples-like

Reply:
I don't know what you are suggesting here

Reply:
In dimod we have a notion of `sample-like`, which can be a lot of things. If we're going to extend from dict to also allow lists/numpy array then we should use the full sample-like notion. See [as_samples](https://docs.ocean.dwavesys.com/projects/dimod/en/latest/reference/generated/dimod.sampleset.as_samples.html#dimod.sampleset.as_samples). So any 'sample-like' object would be a valid input as samples. This would also allow us to do `poly_energies`.

Reply:
I can also change this later